From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Fri, 26 Jan 2024 01:36:34 +0800
Subject: [PATCH] Leaves: Vanilla hopper

Original license: GPL-3.0-only
Original project: https://github.com/LeavesMC/Leaves

This is a temporary solution designed to attempt to restore the vanilla behavior of the funnel while preserving optimizations as much as possible. It should ultimately be replaced by the optimization solution provided by lithium.

diff --git a/net/minecraft/world/level/block/entity/HopperBlockEntity.java b/net/minecraft/world/level/block/entity/HopperBlockEntity.java
index c61355973fcf48e8bb586de8bb852a39f426b750..4e8098cecc173d1ea6e97a6c67e0bbfd1ffdd86a 100644
--- a/net/minecraft/world/level/block/entity/HopperBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/HopperBlockEntity.java
@@ -283,41 +283,77 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
         ItemStack movedItem = origItemStack;
         final int originalItemCount = origItemStack.getCount();
         final int movedItemCount = Math.min(level.spigotConfig.hopperAmount, originalItemCount);
-        movedItem.setCount(movedItemCount);
+        // Leaves start - Vanilla hopper
+        if (org.dreeam.leaf.config.modules.gameplay.VanillaHopper.enabled && movedItemCount == 1) {
+            movedItem.setCount(movedItemCount);
+            if (!skipPullModeEventFire) {
+                movedItem = callPullMoveEvent(hopper, container, movedItem);
+                if (movedItem == null) { // cancelled
+                    origItemStack.setCount(originalItemCount);
+                    // Drastically improve performance by returning true.
+                    // No plugin could have relied on the behavior of false as the other call
+                    // site for IMIE did not exhibit the same behavior
+                    return true;
+                }
+            }
+            final boolean removeOriginalItem = movedItem == origItemStack;
+            if (removeOriginalItem) {
+                movedItem = container.removeItem(i, movedItemCount);
+                if (movedItem == origItemStack) {
+                    movedItem = movedItem.copy(true); // copy if reference equal
+                }
+            }
 
-        if (!skipPullModeEventFire) {
-            movedItem = callPullMoveEvent(hopper, container, movedItem);
-            if (movedItem == null) { // cancelled
-                origItemStack.setCount(originalItemCount);
-                // Drastically improve performance by returning true.
-                // No plugin could have relied on the behavior of false as the other call
-                // site for IMIE did not exhibit the same behavior
+            final int itemCountToMove = movedItem.getCount();
+            final ItemStack remainingItem = addItem(container, hopper, movedItem, null);
+
+            final int remainingItemCount = remainingItem.getCount();
+            if (remainingItemCount != itemCountToMove) {
+                origItemStack.setCount(removeOriginalItem ? originalItemCount - movedItemCount + remainingItemCount : originalItemCount);
+                container.setItem(i, origItemStack);
+                container.setChanged();
                 return true;
             }
-        }
-
-        final ItemStack remainingItem = addItem(container, hopper, movedItem, null);
-        final int remainingItemCount = remainingItem.getCount();
-        if (remainingItemCount != movedItemCount) {
-            origItemStack = origItemStack.copy(true);
             origItemStack.setCount(originalItemCount);
-            if (!origItemStack.isEmpty()) {
-                origItemStack.setCount(originalItemCount - movedItemCount + remainingItemCount);
+            container.setItem(i, origItemStack);
+        } else {
+            movedItem.setCount(movedItemCount);
+
+            if (!skipPullModeEventFire) {
+                movedItem = callPullMoveEvent(hopper, container, movedItem);
+                if (movedItem == null) { // cancelled
+                    origItemStack.setCount(originalItemCount);
+                    // Drastically improve performance by returning true.
+                    // No plugin could have relied on the behavior of false as the other call
+                    // site for IMIE did not exhibit the same behavior
+                    return true;
+                }
             }
 
-            ignoreBlockEntityUpdates = true;
-            container.setItem(i, origItemStack);
-            ignoreBlockEntityUpdates = false;
-            container.setChanged();
-            return true;
-        }
-        origItemStack.setCount(originalItemCount);
+            final ItemStack remainingItem = addItem(container, hopper, movedItem, null);
+            final int remainingItemCount = remainingItem.getCount();
+            if (remainingItemCount != movedItemCount) {
+                origItemStack = origItemStack.copy(true);
+                origItemStack.setCount(originalItemCount);
+                if (!origItemStack.isEmpty()) {
+                    origItemStack.setCount(originalItemCount - movedItemCount + remainingItemCount);
+                }
 
-        Container updatedContainer = container;
-        if (container instanceof net.minecraft.world.CompoundContainer compoundContainer) {
-            updatedContainer = i >= compoundContainer.container1.getContainerSize() ? compoundContainer.container2 : compoundContainer.container1;
+                ignoreBlockEntityUpdates = true;
+                container.setItem(i, origItemStack);
+                ignoreBlockEntityUpdates = false;
+                container.setChanged();
+                return true;
+            }
+            origItemStack.setCount(originalItemCount);
+
+            Container updatedContainer = container;
+            if (container instanceof net.minecraft.world.CompoundContainer compoundContainer) {
+                updatedContainer = i >= compoundContainer.container1.getContainerSize() ? compoundContainer.container2 : compoundContainer.container1;
+            }
+            updatedContainer.setChanged();
         }
-        updatedContainer.setChanged();
+        // Leaves end - Vanilla hopper
 
         if (level.paperConfig().hopper.cooldownWhenFull) {
             applyCooldown(hopper);
